go/benchmarks/benchmarking.go

```diff
@@ -0,0 +1,418 @@
+	}
+
+	return nil
+}
+
+type benchmarkResults struct {
+	iterationsPerTask int
+	duration          time.Duration
+	tps               float64
+	latencyStats      map[string]*latencyStats
+}
+
+func measureBenchmark(clients []benchmarkClient, config *benchmarkConfig) *benchmarkResults {
+	var iterationsPerTask int
+	if config.minimal {
+		iterationsPerTask = 1000
+	} else {
+		iterationsPerTask = int(math.Min(math.Max(1e5, float64(config.numConcurrentTasks*1e4)), 1e7))
+	}
+
+	actions := getActions(config.dataSize)
+	duration, latencies := runBenchmark(iterationsPerTask, config.numConcurrentTasks, actions, clients)
+	tps := calculateTPS(latencies, duration)
+	stats := getLatencyStats(latencies)
+	return &benchmarkResults{
+		iterationsPerTask: iterationsPerTask,
+		duration:          duration,
+		tps:               tps,
+		latencyStats:      stats,
+	}
+}
+
+func calculateTPS(latencies map[string][]time.Duration, totalDuration time.Duration) float64 {
+	numRequests := 0
+	for _, durations := range latencies {
+		numRequests += len(durations)
+	}
+
+	return float64(numRequests) / totalDuration.Seconds()
+}
+
+type operations func(client benchmarkClient) (string, error)
+
+const (
+	getExisting    = "get_existing"
+	getNonExisting = "get_non_existing"
+	set            = "set"
+)
+
+func getActions(dataSize int) map[string]operations {
+	actions := map[string]operations{
+		getExisting: func(client benchmarkClient) (string, error) {
+			return client.get(keyFromExistingKeyspace())
+		},
+		getNonExisting: func(client benchmarkClient) (string, error) {
+			return client.get(keyFromNewKeyspace())
+		},
+		set: func(client benchmarkClient) (string, error) {
+			return client.set(keyFromExistingKeyspace(), strings.Repeat("0", dataSize))
+		},
+	}
+
+	return actions
+}
+
+const (
+	sizeNewKeyspace      = 3750000
+	sizeExistingKeyspace = 3000000
+)
+
+func keyFromExistingKeyspace() string {
+	localRand := rand.New(rand.NewSource(time.Now().UnixNano()))
+	return fmt.Sprint(math.Floor(localRand.Float64()*float64(sizeExistingKeyspace)) + 1)
+}
+
+func keyFromNewKeyspace() string {
+	localRand := rand.New(rand.NewSource(time.Now().UnixNano()))
+	totalRange := sizeNewKeyspace - sizeExistingKeyspace
+	return fmt.Sprint(math.Floor(localRand.Float64()*float64(totalRange) + sizeExistingKeyspace + 1))
+}
+
+type actionLatency struct {
+	action  string
+	latency time.Duration
+}
+
+func runBenchmark(
+	iterationsPerTask int,
+	concurrentTasks int,
+	actions map[string]operations,
+	clients []benchmarkClient,
+) (totalDuration time.Duration, latencies map[string][]time.Duration) {
+	latencies = map[string][]time.Duration{
+		getExisting:    {},
+		getNonExisting: {},
+		set:            {},
+	}
+
+	numResults := concurrentTasks * iterationsPerTask
+	results := make(chan *actionLatency, numResults)
+	start := time.Now()
+	for i := 0; i < concurrentTasks; i++ {
+		go runTask(results, iterationsPerTask, actions, clients)
```

Comment:
Is it an async operation?
Sorry my golang knowledge is bad.

Reply:
It is async. If you put `go` before calling a function it will spawn that function in a new goroutine, meaning the calls to `go runTask` will run concurrently. Because the `results` channel was given a size, it is buffered which means writes to the channels are asynchronous. Writes to the `results` channel will only block if the buffer is full, but the size of the buffer is equal to the number of writes so that won't be possible.